### PR TITLE
Zero Alloc WriteTo

### LINKED ISSUE #805: Bitmap.WriteTo performs unecessary allocations
### Description
The purpose of `Bitmap.WriteTo` is to serialize the `roaring.Bitmap` to an `io.Writer`. Currently it allocates a byte slice, populates the byte slice, and then writes it to the `io.Writer`. It should be possible to directly write bytes to the `io.Writer` rather than going through a byte slice first.

This conversion will involve using the `binary.Write(w io.Writer, order ByteOrder, data interface{}) error` function instead of the `binary.LittleEndian.Put*` methods.

### Success criteria (What criteria will consider this ticket closeable?)
Bitmap.WriteTo should no longer allocate additional memory and performance should be improved.

### PR DESCRIPTION
**Pull request by tgruben**
## Overview

Fixes #805

## Pull request checklist

- [x] I have read the [contributing guide](https://github.com/pilosa/pilosa/blob/master/CONTRIBUTING.md).
- [x] I have agreed to the [Contributor License Agreement](https://cla-assistant.io/pilosa/pilosa).
- [x] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [x] I have resolved any merge conflicts.
- [x] I have included tests that cover my changes.
- [x] All new and existing tests pass.

## Code review checklist
This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.

### CODE REVIEW

File: fragment.go [hunk 1]
```diff
@@ -1262,6 +1263,7 @@ func (f *Fragment) snapshot() error {
 	} else if err := bw.Flush(); err != nil {
 		return fmt.Errorf("flush: %s", err)
 	}
+	bw.Flush()
```
Comment (jaffee) [inline fragment.go]: didn't we just flush right above?

File: roaring/roaring.go [hunk 2]
```diff
@@ -526,15 +541,18 @@ func (b *Bitmap) WriteTo(w io.Writer) (n int64, err error) {
...
 	containerCount := len(b.keys) - b.countEmptyContainers()
 	headerSize := headerBaseSize
+	byte2 := make([]byte, 2)
+	byte4 := make([]byte, 4)
+	byte8 := make([]byte, 8)
 
 	// Build header before writing individual container blocks.
 	// Metadata for each container is 8+2+2+4 = sizeof(key) + sizeof(container_type)+sizeof(cardinality) + sizeof(file offset)
-	buf := make([]byte, headerSize+(containerCount*(8+2+2+4)))
+	//	buf := make([]byte, headerSize+(containerCount*(8+2+2+4)))
```
Comment (jaffee) [inline roaring/roaring.go]: rm

File: roaring/roaring.go [hunk 3]
```diff
@@ -1632,8 +1643,10 @@ func (c *container) runWriteTo(w io.Writer) (n int64, err error) {
 	if len(c.runs) == 0 {
 		return 0, nil
 	}
+	var byte2 [2]byte
 	// Write sizeof(interval16) * runCount bytes.
-	err = binary.Write(w, binary.LittleEndian, uint16(len(c.runs)))
+	//	err = binary.Write(w, binary.LittleEndian, uint16(len(c.runs)))
```
Comment (jaffee) [inline roaring/roaring.go]: rm

File: fragment.go (hunk 1 above)
Comment (travisturner) [inline fragment.go]: this new line `bw.Flush()` seems extraneous. is there a reason to call it again?

somewhat related, i think the `else` statement before this makes it confusing to read. i.e. i think we could remove the `else` and just have a new `if` statement.

File: fragment.go [hunk 4]
```diff
@@ -206,10 +206,11 @@ func (f *Fragment) openStorage() error {
 	if err != nil {
 		return err
 	} else if fi.Size() == 0 {
-		if _, err := f.storage.WriteTo(f.file); err != nil {
+		bi := bufio.NewWriter(f.file)
```
Comment (jaffee) [inline fragment.go]: would it make sense to add the buffered writer inside `f.storage.WriteTo`? ...maybe not, not sure what best practice is here

Comment (tgruben) [reply fragment.go]: My preference to put the implementation of the interface as high as possible in the stack as possible

**Pull request merged**